irods/at_client_exit.py

```diff
@@ -0,0 +1,42 @@
+import atexit
+import enum
+
+_cleanup_functions = {}
+
+class priority(enum.Enum):
+    BEFORE_PRC = 10
+    DURING_PRC = 20
+    AFTER_PRC = 30
+
+# So that we can access priority.ENUM_NAME as simply ENUM_NAME at module level.
+for _ in priority.__members__:
+    exec('{}=priority.{}'.format(_,_))
+
+initialized = False
+
+def register(priority_for_execution, function):
+    if not initialized:
+        initialized = True
+        atexit.register(call_cleanup)
+    _cleanup_functions[function] = priority( priority_for_execution )
+
+def call_cleanup():
+   funclist = sorted((pri.value,func) for func,pri in _cleanup_functions.items())
+   for priority,function in funclist:
+       try:
+           # Ensure we execute all cleanup functions regardless of some of them raising exceptions.
+           function()
+       except Exception as exc:
+           logger.getLogger(__name__).debug("%r raised from %s",exc,function)
```

Comment:
should probably be upgraded to the INFO log level.

Reply:
Why INFO instead of WARNING?

Reply:
WARNING will work.

Reply:
It doesn't have to be warning. I was just wondering if warning is more appropriate.

Reply:
Yeah, I suppose it's the difference between such warnings being logged by default or not. I think there's good arguments for WARNING being used, to avoid masking bad things that could be going on during an application;s shutdown.

Reply:
Seems bad….  I vote warning

Reply:
Let's go with warning.